phantazm-mob/src/main/java/com/github/phantazmnetwork/mob/config/MobModelConfigProcessor.java

```diff
@@ -0,0 +1,165 @@
+
+        ConfigList goalGroupsList = element.getListOrThrow("goalGroups");
+        Collection<Collection<Goal>> goalGroups = new ArrayList<>(goalGroupsList.size());
+        for (ConfigElement goalGroupElement : goalGroupsList) {
+            if (!goalGroupElement.isList()) {
+                throw new ConfigProcessException("goal groups are not a list");
+            }
+
+            ConfigList goalsList = goalGroupElement.asList();
+            Collection<Goal> goals = new ArrayList<>(goalsList.size());
+            for (ConfigElement goalElement : goalsList) {
+                goals.add(goalProcessor.dataFromElement(goalElement));
+            }
+
+            goalGroups.add(List.copyOf(goals));
+        }
+
+        ConfigList triggersList = element.getListOrThrow("triggers");
+        Map<Key, Collection<Skill>> triggers = new HashMap<>(triggersList.size());
+        for (ConfigElement triggerElement : triggersList) {
+            Key triggerKey = KEY_PROCESSOR.dataFromElement(triggerElement.getElementOrThrow("key"));
+
+            ConfigList skillsList = triggerElement.getListOrThrow("skills");
+            Collection<Skill> skills = new ArrayList<>(skillsList.size());
+            for (ConfigElement skillElement : skillsList) {
+                Skill skillInstance = skillProcessor.dataFromElement(skillElement);
+                skills.add(skillInstance);
+            }
+
+            triggers.put(triggerKey, List.copyOf(skills));
+        }
+
+        Component displayName = componentProcessor.dataFromElement(element.getElementOrThrow("displayName"));
+
+        ConfigNode equipmentNode = element.getNodeOrThrow("equipment");
+        Map<EquipmentSlot, ItemStack> equipment = new HashMap<>(equipmentNode.size());
+        for (Map.Entry<String, ConfigElement> entry : equipmentNode.entrySet()) {
+            EquipmentSlot equipmentSlot = EQUIPMENT_SLOT_PROCESSOR.dataFromElement(new ConfigPrimitive(entry.getKey().toUpperCase()));
+            equipment.put(equipmentSlot, itemStackProcessor.dataFromElement(entry.getValue()));
+        }
+
+        float maxHealth = element.getNumberOrThrow("maxHealth").floatValue();
+
+        return new MobModel(key, descriptor, goalGroups, triggers, displayName, equipment, maxHealth);
+    }
+
+    @Override
+    public @NotNull ConfigElement elementFromData(@NotNull MobModel model) throws ConfigProcessException {
+        ConfigElement key = KEY_PROCESSOR.elementFromData(model.key());
+
+        ConfigElement descriptor = descriptorProcessor.elementFromData(model.getDescriptor());
+
+        ConfigList goalGroupsList = new ArrayConfigList(model.getGoalGroups().size());
+        for (Collection<Goal> group : model.getGoalGroups()) {
+            ConfigList goalGroupList = new ArrayConfigList(group.size());
+            for (Goal goal : group) {
+                goalGroupList.add(goalProcessor.elementFromData(goal));
+            }
+
+            goalGroupsList.add(goalGroupList);
```

Comment:
there is not a List.copyOf here if that matters

Reply:
serialization